services/notification/src/helpers/database-helper.js

```diff
@@ -15,19 +16,158 @@ class DatabaseHelper {
     if (cachedDb && cachedDb.serverConfig.isConnected()) {
       this.db = cachedDb;
     } else {
-      const client = await MongoClient.connect(this.uri);
+      const client = await MongoClient.connect(this.uri, {
+        useUnifiedTopology: true,
+      });
       this.db = client.db(this.dbName);
     }
   }
 
   async findNotifications(frequency) {
     if (frequency === EmailFrequency.INSTANT) {
       return this._findInstantNotifications();
+    } else if (frequency === EmailFrequency.MESSAGE) {
+      return this._findUnreadDirectMessages();
     }
     return this._findDigestNotifications(frequency);
   }
 
-  async setEmailSentAt(notificationIds, frequency) {
+  async _findUnreadDirectMessages() {
+    const threads = await this.db
+      .collection("threads")
+      .aggregate([
+        {
+          $match: {
+            "participants.newMessages": { $gt: 0 },
+            "participants.emailSent": false,
+            "participants.status": {
+              $in: [MessageThreadStatus.ACCEPTED, MessageThreadStatus.PENDING],
+            },
+            $or: [
+              {
+                "participants.lastAccess": null,
+              },
+              {
+                $and: [
+                  {
+                    "participants.lastAccess": {
+                      $lt: DateHelper.subtractMinutes(
+                        new Date(),
+                        this.instantUnreadLookbackInterval,
+                      ),
+                    },
+                  },
+                  {
+                    "participants.lastAccess": {
+                      $gt: DateHelper.subtractMinutes(new Date(), 30),
+                    },
+                  },
+                ],
+              },
+            ],
+          },
+        },
+        {
+          $lookup: {
+            from: "users",
+            localField: "participants.id",
+            foreignField: "_id",
+            as: "user",
+          },
+        },
+        {
+          $project: {
+            participants: {
+              $map: {
+                input: "$participants",
+                in: {
+                  id: "$$this.id",
+                  emailSent: "$$this.emailSent",
+                  lastAccess: "$$this.lastAccess",
+                  name: "$$this.name",
+                  newMessages: "$$this.newMessages",
+                  photo: "$$this.photo",
+                  status: "$$this.status",
+                  type: "$$this.type",
+                  user: {
+                    $arrayElemAt: [
+                      "$user",
+                      {
+                        $indexOfArray: ["$user._id", "$$this.id"],
+                      },
+                    ],
+                  },
+                },
+              },
+            },
+          },
+        },
+      ])
+      .toArray();
+
+    const threadIds = threads.map((thread) => thread._id);
+
+    // Convert threads array to object keyed by thread ID for O(1) lookup
```

Comment:
this seems will still be O(n), [ES Spec](https://tc39.es/ecma262/#sec-array.prototype.reduce) description says current value comes from previous result, so there might be no parallel happening

Reply:
Yes, it's `O(n)` when converting from an array to object (which makes sense since it has to process every item in the array), but the reason for converting to an object (or hash lookup) is for `O(1)` lookup when it finds the message by thread ID [here](https://github.com/FightPandemics/FightPandemics/pull/1821/files#diff-63ecab986d8af29fb81344382b42ee27cfb405e5bd945c4d5dc3b70f37d27983R138). So overall `O(n) + O(1) ~ O(n)`. If we left it as an array, yes we save `O(n)` time in converting to an object but the lookup will be `O(n^2)` since it has to look for the element in the array by iterating through every element. So that would be `O(n^2)`. `O(n) << O(n^2)`

Reply:
that makes sense, I misunderstood it ...